### PR TITLE
Improvements to intra-doc link macro disambiguators

### LINKED ISSUE #86309: Intra-doc link `foo!()` is rendered without the parentheses
Consider this code:

```rust
#[macro_export] macro_rules! foo {
    () => {}
}

/// [foo!()]
pub struct Foo;
```

In the generated docs, the link on the `Foo` page appears as `foo!`, with the correct link, but without the parentheses.

### LINKED ISSUE #86310: Intra-doc links don't accept all macro delimiters
Consider this code:

```rust
// This can be invoked with `foo!()`, `foo[]` or `foo{}`
#[macro_export] macro_rules! foo {
    () => {}
}

/// This is parsed as an intra-doc link: [foo!()]...
/// ...but those aren't: [foo![]] [foo!{}]
pub struct Foo;
```

I'm not sure if this is a deliberate decision or not. If it is, what's the rationale for not accepting `[]` and `{}`? cc @jyn514

### PR DESCRIPTION
**Pull request by LeSeulArtichaut**
A few small improvements around macro disambiguators:
- display the link text as it was entered: previously `[macro!()]` would be displayed without the parantheses (fixes #86309)
- support `!{}` and `![]` as macro disambiguators (fixes #86310)

r? @jyn514 cc @Manishearth @camelid

### CODE REVIEW

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 1]
```diff
@@ -951,9 +951,9 @@ fn preprocess_link<'a>(
     }
 
     // Parse and strip the disambiguator from the link, if present.
-    let (path_str, disambiguator) = match Disambiguator::from_str(&link) {
-        Ok(Some((d, path))) => (path.trim(), Some(d)),
-        Ok(None) => (link.trim(), None),
+    let (link_text, path_str, disambiguator) = match Disambiguator::from_str(&link) {
+        Ok(Some((d, path, link_text))) => (link_text, path.trim(), Some(d)),
+        Ok(None) => (link.trim().to_owned(), link.trim(), None),
```
Comment (jyn514) [inline src/librustdoc/passes/collect_intra_doc_links.rs:956]: It seems confusing to have the order of the function different from the order of the match.
```suggestion
    let (disambiguator, path_str, link_text) = match Disambiguator::from_str(&link) {
        Ok(Some((d, path, link_text))) => (Some(d), path.trim(), link_text),
        Ok(None) => (None, link.trim(), link.trim().to_owned()),
```

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 2]
```diff
@@ -1551,21 +1534,29 @@ impl Disambiguator {
...
                     // Avoid turning `!` or `()` into an empty string
-                    if !link.is_empty() {
-                        return Ok(Some((Kind(kind), link)));
+                    if !path_str.is_empty() {
+                        return Ok(Some((Kind(kind), path_str, link.trim().to_owned())));
                     }
                 }
             }
+            // Parse `macro!\[\]` (backslashes are needed in link labels)
+            if let Some(path_str) = link.strip_suffix(r"!\[\]") {
```
Comment (jyn514) [inline src/librustdoc/passes/collect_intra_doc_links.rs:1555]: This seems like a misfeature (it also supports, for example, `[m](m!\[\])` and `[m]: m!\[\]`). I don't think we should adapt the markdown syntax, people will find it surprising that `![]` isn't supported either way and there are plenty of other workarounds.

If you remove this, please also change the return type from String to &str.

**Comment by LeSeulArtichaut** [comment]
@jyn514 Addressed your comments

**Review by jyn514 — APPROVED** [review]

**Comment by jyn514** [comment]
@bors r+

Thanks!

**Pull request merged**